### PR TITLE
Add Mistral models ONNX export support

### PR DESCRIPTION
**Pull request by echarlaix**

### CODE REVIEW

File: tests/onnxruntime/test_modeling.py [hunk 1]
```diff
@@ -2083,7 +2084,7 @@ def test_compare_to_transformers(self, test_name: str, model_arch: str, use_cach
         set_seed(SEED)
         transformers_model = AutoModelForCausalLM.from_pretrained(model_id)
         transformers_model = transformers_model.eval()
-        tokenizer = get_preprocessor(model_id)
+        tokenizer = AutoTokenizer.from_pretrained(model_id, use_fast=False if model_arch == "mistral" else True)
```
[the remark this answers is not in the document]
Comment (echarlaix) [reply tests/onnxruntime/test_modeling.py:2087]: For the tests we use tiny models, here https://huggingface.co/hf-internal-testing/tiny-random-MistralForCausalLM which cannot be loaded without adding `use_fast=False`, opened a PR to fix it https://huggingface.co/hf-internal-testing/tiny-random-MistralForCausalLM/discussions/6 (cc @ydshieh), but will replace it with https://huggingface.co/echarlaix/tiny-random-mistral so that we can merge

File: optimum/onnxruntime/base.py [hunk 2]
```diff
@@ -527,7 +534,7 @@ def compute_past_key_values_output_shapes(
     ) -> Dict[str, int]:
         batch_size = input_ids.size(0)
         num_attention_heads = self.normalized_config.num_attention_heads
-        embed_size_per_head = self.normalized_config.hidden_size // num_attention_heads
+        embed_size_per_head = self.normalized_config.hidden_size // self.normalized_config.num_attention_heads
```
Comment (fxmarty) [inline optimum/onnxruntime/base.py:537]:
```suggestion
        embed_size_per_head = self.normalized_config.hidden_size // num_attention_heads
```

File: optimum/onnxruntime/base.py [hunk 3]
```diff
@@ -277,8 +281,11 @@ def compute_past_key_values_output_shapes(
             `Dict[str, List[int]]`: The dictionary mapping each past key value output name to its corresponding shape.
         """
         batch_size = input_ids.size(0)
-        num_attention_heads = self.normalized_config.num_attention_heads
-        embed_size_per_head = self.normalized_config.hidden_size // num_attention_heads
+        if self.normalized_config.config.model_type in {"mistral", "llama"}:
+            num_attention_heads = self.normalized_config.num_attention_heads
+        else:
+            num_attention_heads = self.normalized_config.num_key_value_heads
+        embed_size_per_head = self.normalized_config.hidden_size // self.normalized_config.num_attention_heads
```
Comment (fxmarty) [inline optimum/onnxruntime/base.py:288]: rather:

```suggestion
        num_attention_heads = self.normalized_config.num_attention_heads
        if self.normalized_config.config.model_type in {"mistral", "llama"}:
            embed_size_per_head = self.normalized_config.hidden_size // self.normalized_config.num_key_value_heads
        else:
            embed_size_per_head = self.normalized_config.hidden_size // num_attention_heads
```

is that not? Referring to https://github.com/huggingface/transformers/blob/87499420bff74122a404d1bb0db2e57e6d1ecfe9/src/transformers/models/llama/modeling_llama.py#L352-L353

File: optimum/onnxruntime/base.py (hunk 2 above)
Comment (fxmarty) [inline optimum/onnxruntime/base.py:537]: same comment as above, may it should be

```python
        if self.normalized_config.config.model_type in {"mistral", "llama"}:
            embed_size_per_head = self.normalized_config.hidden_size // self.normalized_config.num_key_value_heads
        else:
            embed_size_per_head = self.normalized_config.hidden_size // num_attention_heads
```

**Review by fxmarty — COMMENTED** [review]
Thank you!

File: optimum/onnxruntime/base.py (hunk 2 above)
Comment (echarlaix) [reply optimum/onnxruntime/base.py:537]: https://github.com/huggingface/transformers/blob/v4.34.0/src/transformers/models/mistral/modeling_mistral.py#L223-L225

Comment (echarlaix) [reply optimum/onnxruntime/base.py:537]: where :
```python
self.head_dim = config.hidden_size // config.num_attention_heads
self.num_key_value_heads = config.num_key_value_heads
```

File: optimum/onnxruntime/base.py (hunk 3 above)
Comment (echarlaix) [reply optimum/onnxruntime/base.py:288]: here `self.head_dim = config.hidden_size // config.num_attention_heads`
https://github.com/huggingface/transformers/blob/87499420bff74122a404d1bb0db2e57e6d1ecfe9/src/transformers/models/llama/modeling_llama.py#L270

Comment (fxmarty) [reply optimum/onnxruntime/base.py:288]: Yes my bad!

**Comment by echarlaix** [comment]
thanks for catching it @fxmarty, added a fix in https://github.com/huggingface/optimum/pull/1425/commits/4df6252f41367c144a7044dba544c4230cb415a7

File: optimum/onnxruntime/base.py (hunk 2 above)
Comment (fxmarty) [reply optimum/onnxruntime/base.py:537]: @echarlaix Here would it make sense to leave it as `embed_size_per_head = self.normalized_config.hidden_size // num_attention_heads` given that `num_attention_heads = self.normalized_config.num_attention_heads` is defined in the line just above?

Comment (fxmarty) [reply optimum/onnxruntime/base.py:537]: nvm

**Review by fxmarty — APPROVED** [review]
LGTM, left a last comment

File: optimum/onnxruntime/base.py (hunk 2 above)
Comment (echarlaix) [reply optimum/onnxruntime/base.py]: Good catch, added it in https://github.com/huggingface/optimum/pull/1425/commits/3d0ac048d92199b23f736f2794ad970b40f81a36
